### PR TITLE
Auto-differentiable numerical inverse & efficient materialization of BNAF masks

### LINKED ISSUE #176: Autodiff problem with block_neural_autoregressive_flow
Very nice package :-)

While playing around with different optimization objectives I ran into an autodiff issue.
The following always returns exactly zeros, which I think isn't correct. This might be related to the bisection search (I think that's what's used here?) but if the while_loop in there is a problem I would have expected an error, not an incorrect result.

```python
import flowjax.flows
import jax.numpy as jnp
import jax
import numpy as np

flow_key = jax.random.PRNGKey(0)
point = np.random.randn(5)
cotan = np.random.randn(5)

base_dist = flowjax.distributions.Normal(jnp.zeros(5))
flow = flowjax.flows.block_neural_autoregressive_flow(flow_key, base_dist=base_dist, invert=True)

out, pull_grad_fn = jax.vjp(lambda x: flow.bijection.transform_and_log_det(x), point)
pullback = pull_grad_fn((cotan, 1.))
pullback
# (Array([0., 0., 0., 0., 0.], dtype=float32),)
```

### PR DESCRIPTION
**Pull request by noahewolfe**
Here, we make two updates which are particularly relevant for block-neural autoregressive flows:

1. Auto-differentiable numerical inverse via the implicit function theorem. (Fixes #176)

- Using the implicit function theorem (I followed https://arxiv.org/abs/2111.00254 in particular), we define a custom jacobian-vector product for `NumericalInverse` transforms.
- Adds `lineax` as a dependency for memory-efficient computation of the JVP.
- I've tested that this works using the `block_neural_autoregressive_flow` and the default greedy bisection search (in a publication to be on the arxiv in the next week or two).
- I added very light unit tests, which pass.

2. Efficient materialization of block masks with Kronecker products

- I found that it was slow to initialize (e.g., just put into memory---no training or anything) block-neural autoregressive flows with large `nn_block_dim`.
- I traced this to `block_diag_mask` and `block_tril_mask`, and reimplemented these with kronecker products.
- These changes pass the unit tests in `test_masks.py`

Let me know what you think, happy to answer any questions and take any feedback!

### CODE REVIEW

File: flowjax/bijections/block_autoregressive_network.py [hunk 1]
```diff
@@ -122,6 +122,10 @@ def __init__(
                 (1, block_dim),
             ]
 
+            def make_layer(inp):
```
Comment (danielward27) [inline flowjax/bijections/block_autoregressive_network.py]: This change seems to be unused?

**Review by danielward27 — CHANGES_REQUESTED** [review]
Thank you so much again for the contribution! It looks good to me, I've just left some comments with some minor suggestions, let me know what you think.

**Review by danielward27 — APPROVED** [review]

**Comment by noahewolfe** [comment]
Just wanted to follow-up here and record for future reference; the publication for which I developed this feature is now on the arXiv!
https://arxiv.org/abs/2602.20277v1
